### DISCUSSION TITLE
Import mode profile from HFSS to use as source in MEEP

### DISCUSSION THREAD

**Discussion by dalarev** [NanoComp/meep#2218] [Q&A]
Hi,

As the title states, I'm wondering if it's possible to import a mode profile from an HFSS simulation for use as a source in MEEP.

It may be possible to compute the mode profile directly in MEEP - if this is the case, may I ask for a link to the relevant documentation? This is for a circular, corrugated, metallic waveguide which could have several properties that could make it an expensive FDTD computation. 

Is this possible? What exactly would I need from HFSS to achieve this? A 2D cross-section of complex field values, I assume.

Appreciate your thoughts and comments.

Cheers!

**Comment by smartalecH** [comment]
So meep uses MPB to compute modes needed for mode sources. MPB requires all materials to be hermitian (so no loss, gain, metals, etc). It also does not support cylindrical coordinates. 

There is an effort to generalize the current `EigenmodeSource` to allow for arbitrary mode solvers. 

But if you've already solved for the mode profile using something like HFSS, and can export it as a `numpy` array, the meep `Source` object can then take that `numpy` array and use that as the source's spatial profile. It will take care of all the necessary interpolation under the hood. Just be sure to create a source for each field component, to specify the proper size, ensure the mode is sufficiently sampled, etc.

**Comment by oskooi** [comment]
> the meep Source object can then take that numpy array and use that as the source's spatial profile.

The `amp_data` property of the `Source` object simply specifies the amplitude profile of the current source which is not necessarily that useful.

A useful feature that currently does not exist but would be nice to support would be to read in a field profile from a file or Numpy array and then use it to form an **equivalent current** using the procedure described in Section 4.2.1 "Principle of Equivalence" of our book chapter: https://arxiv.org/abs/1301.5366.

**Reply by smartalecH** [reply]
IIRC the equivalent source for a mode is just the mode fields themselves (there may be some weird sign flipping/conjugation due to mode orientation... but I don't remember). This might only be the case for hermitian materials though...

Worth going through it again to see if there are any simplifications.

**Reply by smartalecH** [reply]
No, Im wrong. 

The electric current source depends on the H profile, and the magnetic current source depends on the E profile (with a -1 sign flip). There's some component direction switching due to the cross product, but luckily the mode profiles are already transverse. 

https://github.com/NanoComp/meep/blob/da552960eb89573e53698233ac5a0f76af7dfc81/src/mpb.cpp#L858-L862

Like @oskooi said, we should add some functionality that automatically computes the equivalent current source from this. It should be rather straightforward to implement...

(Right now, @dalarev, you can just do this manually)

**Reply by dalarev** [reply]
Very cool - thanks for the info!

I assume regardless of what mode you are sourcing (TE, TM, or some hybrid HE mode), both complex E and H field profiles are required from HFSS?

**Reply by smartalecH** [reply]
Yes, that's right. 

I would start with something simple, like a dielectric waveguide in 2D. Then work your way up to your actual problem. You can timestep for just a little bit, and then plot the fields (eg using `plot2D()`) to ensure the mode is propagating as intended. Feel free to post updates on this discussion.

**Reply by smartalecH** [reply]
Cc @hammy4815

**Reply by smartalecH** [reply]
@hammy4815 you wrote some code that does this right? Care to share?

**Reply by ianmatthewhammond** [reply]
Yes @smartalecH ,

I wrote a very specific case for GaussianBeam2D (which will soon get merged once some other components are complete). I extended the functionality into a generalized function, which I will show below, however, I have not yet tested it beyond the case of 2D. It should be clear what it's doing and can be changed as needed. Eventually meep will add native support for this, but for beyond just single frequency. Here is my function:

Also note, this will always propagate in the positive direction of whatever the normal is. If you want to flip it, I believe you can just flip the sign of nHat.

```python
def get_equiv_sources(field:np.ndarray, src: mp.SourceTime, center: mp.Vector3, size: mp.Vector3):
    """
    + **`field` [np.ndarray]** the combined fields Ex, Ey, Ez, Hx, Hy, Hz of shape 
        (6, Nx, Ny, Nz) for 3D and (6, Nx, Ny, 1) for 2D. Note this is a plane. Thus, 
        for 3D, only two axes should have more than one point, while 2D should only have one
        axis with more than 1 point. No default.
   
    + **`src` [`SourceTime` class ]** — Specify the time-dependence of the source. No default.

    + **`center` [`Vector3`]** — The location of the center of the current source in
        the cell. No default.

    + **`size` [`Vector3`]** — The size of the current distribution along each
        direction of the cell. Represents the equivalent source plane and should be the plane 
        the fields are calculated at. No default.
    """

    # Get fields
    Ex, Ey, Ez, Hx, Hy, Hz = field
    
    # Get normal vector nHat
    if size[2]: # 3D sim
        if size[0] and size[1]:
            raise Exception("size should be a plane and thus have 2 non-zero values in 3D")
        elif size[0]:
            nHat = mp.Vector3(0, 1, 0)
        elif size[1]:
            nHat = mp.Vector3(1, 0, 0)
        else:
            raise Exception("size should be a plane and thus have 2 non-zero values in 3D")
    elif size[0] and size[1]: # 3D sim
        nHat = mp.Vector3(0, 0, 1)
    else: # 2D 
        if size[0]:
            nHat = mp.Vector3(0, 1)
        elif size[1]:
            nHat = mp.Vector3(1, 0)
        else:
            raise Exception("size should be a plane and thus have 1 non-zero values in 2D")

    # Electric current K = nHat x H
    Kx = nHat[1] * Hz - nHat[2] * Hy
    Ky = nHat[2] * Hx - nHat[0] * Hz
    Kz = nHat[0] * Hy - nHat[1] * Hx

    # Mangnetic current N = - nHat x E
    Nx = nHat[2] * Ey - nHat[1] * Ez 
    Ny = nHat[0] * Ez - nHat[2] * Ex 
    Nz = nHat[1] * Ex - nHat[0] * Ey 

    # Source components 
    components = {
        mp.Ex: Kx,
        mp.Ey: Ky,
        mp.Ez: Kz,
        mp.Hx: Nx,
        mp.Hy: Ny,
        mp.Hz: Nz
    }

    # Make sources
    sources = [
        mp.Source(src, field_comp, center=center, size=size, amp_data=source_comp) 
        for field_comp, source_comp in components.items() if np.sum(np.abs(source_comp)) 
    ]

    return sources
```

**Reply by dalarev** [reply]
Hey @hammy4815 thanks for the function. 

I'm working through it but am having issues with the size of `field` and what `src` should be. Is there some test code somewhere that could be used as reference?

For example, should component `Ex` be a complex vector of length `Nx = Ny = Nz`?

I'll keep working through this. Thanks!

**Reply by smartalecH** [reply]
Did you look at the function docstring?

```python
"""
+ **`field` [np.ndarray]** the combined fields Ex, Ey, Ez, Hx, Hy, Hz of shape 
        (6, Nx, Ny, Nz) for 3D and (6, Nx, Ny, 1) for 2D. Note this is a plane. Thus, 
        for 3D, only two axes should have more than one point, while 2D should only have one
        axis with more than 1 point. No default.
   
    + **`src` [`SourceTime` class ]** — Specify the time-dependence of the source. No default.
"""
```

Do you have a specific question about that?

> For example, should component Ex be a complex vector of length Nx = Ny = Nz?

As mentioned in the docstring, one of the array (not vector) axes must be of length 1 (in 3d), otherwise it's not a plane. In 2d, 2 axes must be of length 1 (one of which must be `nz`). There's no other requirement forcing them to equal each other. 

I think you're misinterpreting dimensions of arrays in numpy. It's a 4D array, not a 2D matrix of vectors.

**Reply by dalarev** [reply]
Re: `field`: I understand now. I was confused because the format my field data is in will require some parsing, but I understand the docstring now. Thanks for your help.

Re: `src`: From what I can tell from the MEEP docs, this will be one of either `ContinuousSource`, `GaussianSource`, or `CustomSource`.

**Reply by ianmatthewhammond** [reply]
@dalarev ,

Yep, that's correct. If you continue to have trouble formatting the `field` parameter, you can take a look at the script I'm attaching here as a working example. It uses the get_equiv_sources function to create equivalent sources for the analytic solution to a 2-D gaussian beam and then runs meep with them. 

[eq-sources-example.py.zip](https://github.com/NanoComp/meep/files/9944455/eq-sources-example.py.zip)

**Reply by dalarev** [reply]
Thanks @hammy4815.

I was able to get some sims running using an .fld file from a dipole in HFSS, but I'm seeing some discrepancies.

I need a bit more time to review my code. Yours will be helpful for comparisons.

**Reply by dalarev** [reply]
Question about `fields` and the size of the field components that fill it.

My source data is in list format, so I use `griddata` to interpolate it onto a regular grid. This uses `np.meshgrid`, which, as you know, returns a `(row, column)` size rather than `(Nx, Ny)`.

So, when I initialize `field = np.ndarray(shape=(6,Nx,Ny,1))`, should I fill in the field components by using the transpose? For example, `field[0,:,:,0] = Ex.T`, where `Ex` is my interpolated field component? I've tried this and a few other combinations, but there appears to be some strange orientation error in my code.

I hope this makes sense. The full script is [here](https://github.com/dalarev/MEEP_scripts/blob/main/test_equivalent_source_SIWave-to-HFSS.ipynb), if that helps.

Appreciate the guidance!

edit: actually, I think the one-line fill was incorrect, so I changed it to a loop where I flip the `i,j` of the field component:
```
for i in range(Nx):
     for j in range(Ny):
          field[0,i,j,0] = Ex[j,i]
          ...
```

But I still think there's some small mistake somewhere because I'm not quite matching the HFSS results yet.

**Reply by ianmatthewhammond** [reply]
Hi @dalarev ,

The way you are filling your array and passing into the eq_source function looks correct to me. 

Do you have enough points? You seem to be interpolating onto a 500x500 point grid from only about 2000 datapoints. I'm not sure if this is the issue however, considering your points seem to be chosen at the high information points in your field.

You seem to be looking at time domain fields, I wonder if you try looking at the meep DFT fields at your frequency point over a long enough simulation if it might look more like your starting DFT fields that you provided to eq_source.

**Reply by dalarev** [reply]
Thanks for the guidance, @hammy4815.

It's easy enough to try a higher-res interpolation grid, so I will give that a shot. I did notice that this parameter changed the results somewhat, so it needs further looking into.

Good idea re: the DFT fields! It is true that I am only interested in results at one frequency, so this makes more sense. I will refer to the docs for guidance on how to use these functions.

I very much appreciate the guidance!

**Comment by stevengj** [comment]
> I assume regardless of what mode you are sourcing (TE, TM, or some hybrid HE mode), both complex E and H field profiles are required from HFSS?

Technically, you can use just the E fields (to make a magnetic current –n×E) or just the H fields (to make an electric current n×H).  This still launches the waveguide mode, but the mode will propagate in both directions away from the source instead of just forwards.   (It corresponds to taking the one-way source and averaging it with ± its mirror flip).

**Reply by dalarev** [reply]
Yes, that's exactly what I'm seeing, but I wasn't 100% on the math behind it - thank you!

**Reply by stevengj** [reply]
See the section on the principal of equivalence in https://arxiv.org/abs/1301.5366

**Comment by dalarev** [comment]
Hi everybody, hoping to get your input on an intriguing result using `get_equiv_sources`. I won't spam code but can provide it as needed.

Trying to validate MEEP `get_equiv_sources` fields with those of a dipole in HFSS. I export near-fields from HFSS and use `get_equiv_sources` to calculate equivalent currents to define the MEEP source. Long story short: when using `get_equiv_sources`, MEEP matches the HFSS field profile in shape but underestimates the peak magnitudes; however, when using the HFSS fields directly, without calculating equivalent currents, the amplitudes more closely match up.

This is summarized in the image below. Again, I don't want to spam with detail, but I'm hoping to get any thoughts you may have with this information. Appreciate your input in advance!

![dipole_validation](https://user-images.githubusercontent.com/52722332/207423920-2aed812f-d01b-48cf-aae3-d0e53fad4975.png)
